Add `padding="same"` for transposed convolution

Alvaro-Kothe opened the pull request:
This pull requests makes `ConvTranspose*d` and `conv_transpose*d` compatible with the argument `padding="same"`.
I tried to follow the current implementation of the `Conv*d` layer.


Closes #80301, Closes #3867 


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

Alvaro-Kothe committed:
feat(ConvTransposeNd): add `padding="same"` for `ConvTranspose*d`

Alvaro-Kothe committed:
docs(ConvTransposeNd): add documentation for `padding="same"`

Alvaro-Kothe committed:
test(conv_transpose*d): add tests for `padding="same"`

Alvaro-Kothe commented:
@pytorchbot label "release notes: nn"

Alvaro-Kothe committed:
fix(ConvTranspose): handle `right_pad > left_pad`

Skylion007 commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -209,6 +209,33 @@ inline void convolution_shape_check(
+    // The padding on the left is ceil(total_padding / 2)
+    left_pad = (total_padding / 2) + (total_padding % 2);
+  }
+  return {left_pad, total_padding - left_pad};
```
May want to forward these args

Skylion007 commented on torch/nn/modules/conv.py:
```diff
@@ -1112,7 +1128,7 @@ def __init__(
         kernel_size: _size_2_t,
         stride: _size_2_t = 1,
-        padding: _size_2_t = 0,
+        padding: Union[str, _size_2_t] = 0,
```
```suggestion
        padding: Union[Literal["Padding", "same"], _size_2_t] = 0,
```

Skylion007 commented on torch/nn/modules/conv.py:
```diff
@@ -922,7 +930,7 @@ def __init__(
         kernel_size: _size_1_t,
         stride: _size_1_t = 1,
-        padding: _size_1_t = 0,
+        padding: Union[str, _size_1_t] = 0,
```
```suggestion
        padding: Union[Literal["valid", "same"], _size_2_t] = 0,

Alvaro-Kothe committed:
fix(ConvTransposeNd): use `Literal` for `padding` types

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

Alvaro-Kothe committed:
Update aten/src/ATen/native/ConvUtils.h

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

Alvaro-Kothe committed:
docs(ConvTransposeNd): use double backtick for inline code

Alvaro-Kothe committed:
fix(conv_transpose): use correct function name in `batchify`

Alvaro-Kothe committed:
format tests

Alvaro-Kothe committed:
fix: fix clang compile error for complex_conv_transpose_same

Tested the compilation in `g++` version 14 and `clang` version 15.

Alvaro-Kothe committed:
test(conv_transpose3d): add tolerance for non-determinism of 1e-5

Alvaro-Kothe commented:
I did the following to fix the CI errors:

- clang build error
	- Made `complex_convolution_transpose_mode` static
	- Tested the build with `clang-15` and `gcc-14` and the convolution tests passed for both.
-  Cuda error in `test_conv_transpose3d_valid_padding_backward_cuda_complex128`
	-  added `nondet_tol=1e-5` as a workaround.
-  XPU error
	-  Rebased the PR.

mikaylagawarecki commented:
Hi @Alvaro-Kothe, thank you very much for sending this PR. Unfortunately, I don't have the bandwidth to review this at the moment but I will get back to it when I do, I'm sorry about that

Alvaro-Kothe committed:
chore: update misleading comment

Asymmetric convolutions slice the extra value on the left.

Alvaro-Kothe committed:
docs(functional): update padding documentation for `conv_transpose*d`

Document about padding="same" and padding="valid".

Alvaro-Kothe commented:
Hi @mikaylagawarecki, I've also made some changes in the C++ API to ensure parity with the python API. Should I include these changes in this PR?

Alvaro-Kothe commented:
@mikaylagawarecki just a soft ping to see if you’ve had a chance to look at my question above :blush:

Alvaro-Kothe commented:
I pushed my C++ changes to #154279. I also think it's best to focus on the C++ API first, as that will simplify the changes made here. I’m converting this into a draft and will request another review after the other PR is merged.
